menuconfig.py

```diff
@@ -299,6 +299,173 @@
+    return max(0, min(idx, 23))
+
+def _rgb_from_gray_idx(c):
+    """
+    Convert a grayscale index to its closet single RGB component
+    """
+    g = c * 10 + 8
+    return (g, g, g)
+
+# Obscure Python: rgb2index is initialized from a reference to a global {}
+# Modification to this dictionary are retained between calls rgb_alloc(), thus
+# making rgb2index a static variable
+def _rgb_alloc(rgb, rgb2index={}):
+    """
+    Initialize a new entry in the xterm palette to the given RGB color, returning its
+    index. If the color has already been initialized, the index of the existing entry is
+    returned.
+
+    Ncurses doesn't allow you to define new colors -- you are allowed only
+    to overwrite existing ones.
+
+    The colors from 0-15 are user-defined and there's no way to query the
+    RGB values so we better leave them untouched.
+
+    The RGB values of colors from 16-255 can be easily calculated. However,
+    colors from 232-255 use a different formula (grayscale) so stick with
+    colors from 16-231. 200+ colors should be plentiful.
+    """
+    if rgb in rgb2index:
+        return rgb2index[rgb]
+
+    # The first 16 colors are user defined and we should not change their
+    # values
+    cn = len(rgb2index) + 16
+    if cn >= curses.COLORS:
+        _warn("Unable to allocate new RGB color.", rgb)
+        return 0
+
+    # Map each RGB component from the range 0-255 to the range 0-1000, which
+    # is what curses uses
+    curses.init_color(cn, *(int(round(x * 1000 / 255)) for x in rgb))
+    rgb2index[rgb] = cn
+
+    return cn
+
+def _color_get(num):
+    """
+    Returns the index of a color that looks like color 'num' in the xterm
+    256-color palette.
+
+    If ncurses supports color changes, We can't return 'num' directly when
+    redefining colors, since we might have overwritten the palette entry at
+    'num'. Instead, we allocate a new color for it, emulating the 256-color
+    palette.
+    """
+
+    # Simplest case -- terminal doesn't support changing the definition of
+    # colors. _color_get_rgb() won't be changing the current palettte
+    # so we can return the color as-is
+    if not curses.can_change_color():
+        return num
+
+    # Standard colors, rgb_alloc() doesn't touch these so
+    # we can return them as-is
+    if num < 16 or num >= 232:
+        return num
+
+    # We're in RGB mode. rgb_alloc() will redefine the colors
+    # from 16-231. Since the RGB value of the "standard" 256 color
+    # palette can be easily calculated, we can emulate them in RGB
+    # mode rather easily
+    num -= 16
+    return rgb_alloc(_rgb_from_6cube((num / 36) % 6, (num / 6) % 6, num % 6))
+
+def _color_get_rgb(rgb):
+    # Lambda for calculating the Euclidean distance between two RGB colors
+    dist = lambda r1, r2: sum((x - y)**2 for x, y in zip(r1, r2))
+
+    # Best case -- terminal supports the changing of colors
+    if curses.COLORS >= 256 and curses.can_change_color():
```

Comment:
Maybe the `if curses.COLORS >= 256` could be factored out:

```python
if curses.COLORS >= 256:
    ...
    if curses.can_change_color():
        ...
    else:
        ...
```

Reply:
Maybe we can change it to curses.COLORS >= 16, as the algorithm assumes we should not change the first 16 colors.

Reply:
Good catch at the division error. Fixed in the new version.

Reply:
> Maybe we can change it to curses.COLORS >= 16, as the algorithm assumes we should not change the first 16 colors.

Hmm... maybe a higher limit would be nice though, to avoid running out of palette entries. Wonder if you ever run across any terminals with 16 < COLORS < 256.